### PR TITLE
Add MNIST test set evaluation and change hyperparameters.

### PR DESCRIPTION
**Pull request by kamalkraj**
- Add MNIST test set evaluation.
- The batch size and the optimizer are changed to `128` and `Adam`, respectively, based on test set evaluation results. Tested with higher batch sizes [256, 512], but found no improvement in performance.

### CODE REVIEW

File: MNIST/main.swift [hunk 1]
```diff
@@ -70,49 +70,76 @@ struct Classifier: Layer {
     }
 }
 
-let epochCount = 12
-let batchSize = 100
+let epochCount: Int = 12
+let batchSize: Int = 128
```
Comment (rxwei) [inline MNIST/main.swift]: These type annotations are not necessary since integer literals infer to Int.

File: MNIST/main.swift [hunk 2]
```diff
@@ -70,49 +70,76 @@ struct Classifier: Layer {
...
 func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
     let start = index * batchSize
     return x[start..<start+batchSize]
 }
 
-let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+let (train_images, train_numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
                                         labelsFile: "train-labels-idx1-ubyte")
-let labels = Tensor<Float>(oneHotAtIndices: numericLabels, depth: 10)
+let train_labels = Tensor<Float>(oneHotAtIndices: train_numericLabels, depth: 10)
```
Comment (rxwei) [inline MNIST/main.swift]: Please use camel case for all Swift names.

File: MNIST/main.swift [hunk 3]
```diff
@@ -70,49 +70,76 @@ struct Classifier: Layer {
...
           [Epoch \(epoch)] \
-          Loss: \(totalLoss), \
-          Accuracy: \(correctGuessCount)/\(totalGuessCount) (\(accuracy))
+          Training Loss: \(train_stats.totalLoss), \
+          Training Accuracy: \(train_stats.correctGuessCount)/\(train_stats.totalGuessCount) (\(train_accuracy)), \
+          Test Loss: \(test_stats.totalLoss), \
+          Test Accuracy: \(test_stats.correctGuessCount)/\(test_stats.totalGuessCount) (\(test_accuracy))
           """)
-}
+}
```
Comment (rxwei) [inline MNIST/main.swift]: Please leave an empty line at the end of the file.

File: MNIST/main.swift (hunk 1 above)
Comment (kamalkraj) [reply MNIST/main.swift]: Updated

File: MNIST/main.swift (hunk 2 above)
Comment (kamalkraj) [reply MNIST/main.swift]: converted to camel case

File: MNIST/main.swift (hunk 3 above)
Comment (kamalkraj) [reply MNIST/main.swift]: added empty line

File: MNIST/main.swift [hunk 4]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
 }
 
-let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+let (trainImages, trainNumericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
                                         labelsFile: "train-labels-idx1-ubyte")
-let labels = Tensor<Float>(oneHotAtIndices: numericLabels, depth: 10)
+let trainLabels = Tensor<Float>(oneHotAtIndices: trainNumericLabels, depth: 10)
+
+let (testImages, testNumericLabels) = readMNIST(imagesFile: "t10k-images-idx3-ubyte",
+                                        labelsFile: "t10k-labels-idx1-ubyte")
```
Comment (rxwei) [inline MNIST/main.swift]: Fix argument alignment here too.

File: MNIST/main.swift [hunk 5]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+let batchSize = 128
 
 func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
     let start = index * batchSize
     return x[start..<start+batchSize]
 }
 
-let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+let (trainImages, trainNumericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
                                         labelsFile: "train-labels-idx1-ubyte")
```
Comment (rxwei) [inline MNIST/main.swift]: Please align the argument label with the one on the previous line.

File: MNIST/main.swift [hunk 6]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+let testLabels = Tensor<Float>(oneHotAtIndices: testNumericLabels, depth: 10)
 
 var classifier = Classifier()
-let optimizer = RMSProp(for: classifier)
+
+let optimizer = Adam(for: classifier)
 
 print("Beginning training...")
 
+struct Stats {
```
Comment (rxwei) [inline MNIST/main.swift]: The [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/#use-terminology-well) generally discourages the use of abbreviations in type names unless it's a term of art. Although `Stats` is easily understood, `Statistics` is the term of art here.

```suggestion
struct Statistics {
```

File: MNIST/main.swift [hunk 7]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+
 // The training loop.
 for epoch in 1...epochCount {
-    var correctGuessCount = 0
-    var totalGuessCount = 0
-    var totalLoss: Float = 0
-    for i in 0 ..< Int(labels.shape[0]) / batchSize {
-        let x = minibatch(in: images, at: i)
-        let y = minibatch(in: numericLabels, at: i)
+    var trainStats = Stats()
```
Comment (rxwei) [inline MNIST/main.swift]:
```suggestion
    var trainStats = Statistics()
```

File: MNIST/main.swift [hunk 8]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+    var testStats = Stats()
```
Comment (rxwei) [inline MNIST/main.swift]:
```suggestion
    var testStats = Statistics()
```

File: MNIST/main.swift [hunk 9]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+let trainLabels = Tensor<Float>(oneHotAtIndices: trainNumericLabels, depth: 10)
+
+let (testImages, testNumericLabels) = readMNIST(imagesFile: "t10k-images-idx3-ubyte",
+                                        labelsFile: "t10k-labels-idx1-ubyte")
+let testLabels = Tensor<Float>(oneHotAtIndices: testNumericLabels, depth: 10)
 
 var classifier = Classifier()
-let optimizer = RMSProp(for: classifier)
+
+let optimizer = Adam(for: classifier)
```
Comment (rxwei) [inline MNIST/main.swift]: Could you please elaborate on why the optimizer was changed and what difference it made? The PR comment didn't describe this change.

File: MNIST/main.swift [hunk 10]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
...
+    }
+
+    let trainAccuracy = Float(trainStats.correctGuessCount) / Float(trainStats.totalGuessCount)
+    let testAccuracy = Float(testStats.correctGuessCount) / Float(testStats.totalGuessCount)
     print("""
           [Epoch \(epoch)] \
-          Loss: \(totalLoss), \
-          Accuracy: \(correctGuessCount)/\(totalGuessCount) (\(accuracy))
+          Training Loss: \(trainStats.totalLoss), \
+          Training Accuracy: \(trainStats.correctGuessCount)/\(trainStats.totalGuessCount) (\(trainAccuracy)), \
```
Comment (rxwei) [inline MNIST/main.swift]: In Swift for TensorFlow repositories, we fit all Swift source code within 100 columns. Could you please make appropriate changes to make it fit?

File: MNIST/main.swift [hunk 11]
```diff
@@ -71,48 +71,75 @@ struct Classifier: Layer {
 }
 
 let epochCount = 12
-let batchSize = 100
+let batchSize = 128
```
Comment (rxwei) [inline MNIST/main.swift]: This change was not described in the PR comment. Could you elaborate on it?

File: MNIST/main.swift (hunk 5 above)
Comment (rxwei) [reply MNIST/main.swift]: This should actually be aligned with the previous _argument_, i.e. `imagesFile: ...`, not the function name.

File: MNIST/main.swift (hunk 9 above)
Comment (kamalkraj) [reply MNIST/main.swift]: Updated

File: MNIST/main.swift (hunk 11 above)
Comment (kamalkraj) [reply MNIST/main.swift]: Updated

File: MNIST/main.swift (hunk 5 above)
Comment (rxwei) [reply MNIST/main.swift]: Could you please adjust the alignment according to my second comment? Thanks!

File: MNIST/main.swift [hunk 12]
```diff
@@ -71,48 +71,78 @@ struct Classifier: Layer {
...
 }
 
-let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+let (trainImages, trainNumericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
                                         labelsFile: "train-labels-idx1-ubyte")
-let labels = Tensor<Float>(oneHotAtIndices: numericLabels, depth: 10)
+let trainLabels = Tensor<Float>(oneHotAtIndices: trainNumericLabels, depth: 10)
+
+let (testImages, testNumericLabels) = readMNIST(imagesFile: "t10k-images-idx3-ubyte",
+                                      labelsFile: "t10k-labels-idx1-ubyte")
```
Comment (rxwei) [inline MNIST/main.swift]:
```suggestion
                                                labelsFile: "t10k-labels-idx1-ubyte")
```

File: MNIST/main.swift [hunk 13]
```diff
@@ -71,48 +71,78 @@ struct Classifier: Layer {
...
+let batchSize = 128
 
 func minibatch<Scalar>(in x: Tensor<Scalar>, at index: Int) -> Tensor<Scalar> {
     let start = index * batchSize
     return x[start..<start+batchSize]
 }
 
-let (images, numericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
+let (trainImages, trainNumericLabels) = readMNIST(imagesFile: "train-images-idx3-ubyte",
                                         labelsFile: "train-labels-idx1-ubyte")
```
Comment (rxwei) [inline MNIST/main.swift]:
```suggestion
                                                  labelsFile: "train-labels-idx1-ubyte")
```

**Comment by rxwei** [comment]
Thank you!
